### ISSUE TITLE
[QUESTION] How to retrieve AccessToken/IDToken By Cognito when a user has logged in via FederatedSignIn?

### ISSUE THREAD

**Issue by kyeljmd** [aws/aws-amplify#703]
Hello,

I've set up an application where users can register to my site/web application. Where 

A.) A User can either join the site via registering their email/password combination and these users will be registered inside a Cognito User Pool.

B.)  A User can join the site via logging in with Google/Facebook. 

Bullet point A works as as expected. The user will submit their username and password, they will be able to login and they will receive the ID Token, AccessToken, and other information necessary as a response(I'm using aws amplify where it supeseded cognito javascript libraries). Now, when these users access a protected resource on API Gateway that has a cognito_authorizer enabled  they will simply pass in 

"Bearer <Token>" And they will be able to access

Where the <Token> is ID Token.

Now, for case B.

I am now able to login via social provider. I was able to configure all the necessary configurations in both google developer console, and registered google as federated entities.

Now, Thanks to aws amplify, I can perform a federated sign in by passing in the id token, and expires_at value that I have received from the google login. 

However the only values I am able to received as a response from cognito is the CognitoIdentityCredentials
(https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CognitoIdentityCredentials.html)

There is no access token, refresh token, and other necessary information. 

Question is. Is it possible for federated sign in to retrieve an access token, and idtoken that is generated by cognito, and use that as token to be passed in as a header whenever I perform a request where a resource has a cognito_authorizer for users who joined my site via social login? Or am Missing certain steps to perform federated login that will return idtoken,and access token which is generated by cognito?

**Comment by mlabieniec** [comment]
@kyeljmd the federated/social support right now in amplify is using Cognito Federated Identities (not user pools). Basically, your cognito user pool is an IDP (identity provider) on a Cognito Federated Identities pool, just the same as a facebook, google etc. provider. So, they are not linked in anyway, when you federate with Cognito Federated Identities you don't get back jwt tokens, you get an identity ID. When you login with User Pools (essentially a separate service) you get back jwt tokens. 

The way you would federate with Cognito User Pools, is by using a hosted UI/OAuth settings:
https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-app-settings.html

This would take your user to a customized hosted ui where they can federate with a social provider (or saml) and it will all send back JWT tokens from user pools which amplify will consume as usual. This feature is currently in beta will be published in the next release:
https://aws.github.io/aws-amplify/media/authentication_guide#setup-your-cognito-app-client

You can install/try the beta with npm:

npm install aws-amplify@beta

**Comment by guilhermedecampo** [comment]
@mlabieniec thank you for the explanations here.

There is any workaround to add the facebook identity provider data into user pool `by hand`?
I have a super customized user flow and can't use the UI you guys already have developed.

Any help is super welcomed.

**Comment by kyeljmd** [comment]
@mlabieniec  Thanks for the prompt response!

Now with that being said, if I were to use the aws amplify beta for testing out the cognito login via federated users/social users. will I be able to retrieve the access_tokens, and id tokens?

**Comment by mlabieniec** [comment]
@kyeljmd yes that's correct, when the hosted UI returns, it will either return a code or all the tokens (based on your config: 'code' or 'token' grant). If token, the jwt's will come on the URL and amplify will inject them into Auth per usual. If code, a code is sent back and amplify requests the tokens for you. After that you can retrieve them per usual: https://aws.github.io/aws-amplify/media/authentication_guide#retrieve-current-session

@guilhermedecampo at this time i do not believe there is a way to do that with Cognito User Pools. @yuntuowang do you know if there is a way to do this currently i.e. federate without the hosted UI in user pools? Something in your road map?

**Comment by kyeljmd** [comment]
@mlabieniec  I've tried the the documentation that you have specified in this thread 

https://aws.github.io/aws-amplify/media/authentication_guide#setup-your-cognito-app-client


Now, I am receiving a response called invalid client id. Here's the sample code I've used.

My question is what is the clientId that it needs? also how would AWS amplify be able to retrieve idToken, and AuthToken?


Here's the code


```
    const config = Auth.configure(environment.amplify);
    const { 
      domain,
      redirectSignIn, 
      redirectSignOut,
      responseType } = config.oauth;
    
    const clientId = config.userPoolWebClientId;
    const url = 'https://' + domain + '/login?redirect_uri=' + redirectSignIn + '&response_type=' + responseType + '&client_id=' + "661927252789-p4nl3er0uoogac9gj0q8mpbpt2vjdum1.apps.googleusercontent.com";
    console.log(clientId);
    window.location.assign(url);

```

**Comment by kyeljmd** [comment]
I was able to run it properly looks like there was a need for me to reconfigure my api credentials on the google developer console. I needed to Generate a credential ID and select oauth2. 

Upon performing a successful login, I was redirected to my configured redirect uri, and it has the id token, and access_token as part of the request parameters.

so I was redirected to  something this

'localhost:4200/myUri#accessToken=<longvalues>&id_token=<somevalues>&token_type=Bearer&expires_in=3600'

However upon using the id_token, and passing it as a request header in one of my api endpoints that has a cognito_authorizer enabled on the API gateway I always end up receiving this error. as a response from api gateway.


> 
<id_token>not a valid key=value pair (missing equal-sign) in Authorization header: 'Bearer <id_token>


My question is, the id_token part of the response the generated by cognito, or no? if not what should be the next steps?

**Comment by yuntuowang** [comment]
Hi @mlabieniec, federate without the hosted UI in user pools is not currently available. We are still discussing about this feature request.

**Comment by guilhermedecampo** [comment]
@yuntuowang can you provide a link to where the discussion is happening?
Thank you so much

**Comment by yuntuowang** [comment]
Hi @guilhermedecampo, there is no link about discussion. This feature request is in our backlog queue. Thanks!

**Comment by dylan-westbury** [comment]
If identityId was available in the claims (like the sub / userId is), for a proxy lambda. All my problems would go away.

**Comment by sksaranraj** [comment]
Hi @yuntuowang 

Any update on receiving the JWT tokens without UI.

**Comment by ssanchezsv** [comment]
Can someone help me to implement with example code **SingIn** & **SignUp** using **Facebook and Google** as Federated Identity with **Cognito AWS**? I need to make a custom login system using that AWS service.

**Comment by schimini** [comment]
I'm pretty sure this issue is still relevant
